### ISSUE TITLE
Guessing the seed hyperparameter

### ISSUE THREAD

**Issue by vrn25** [alexis-jacq/LOLA_DiCE#4]
https://github.com/alexis-jacq/LOLA_DiCE/blob/ec3f3f620a67df0a3e72f8e9227d09e4543dbb99/ipd_DiCE.py#L12-L22
@alexis-jacq Thanks a lot for such a readable implementation!
I ran LOLA-DiCE for num_lookaheads = 1 for 300 iterations with different seeds (the rest of the hyperparameters were the same). The ones used in the experiments in this repository correspond to seed = 42. I generated the seed using `random.randint(0,100)` function and it showed the following results (the figures show lookahead = 0 in the legend, but that is a typo - it is actually lookahead = 1):

`seed 42`
![seed42](https://user-images.githubusercontent.com/43922602/103468976-eea8e500-4d84-11eb-9217-2a18ae6d585f.png)
`seed 15`
![seed15](https://user-images.githubusercontent.com/43922602/103468983-02ece200-4d85-11eb-86f3-9ebb0e76906a.png)
`seed 45`
![seed45](https://user-images.githubusercontent.com/43922602/103468984-0718ff80-4d85-11eb-8ff2-a23e063b8cda.png)
`seed 64`
![seed64](https://user-images.githubusercontent.com/43922602/103468986-0d0ee080-4d85-11eb-9ac2-3029393df3cf.png)
The seed in each of the experiments is generated using the `random.randint(0,100)` function.

Following is the result when I simply remove seed as a hyperparameter in 
https://github.com/alexis-jacq/LOLA_DiCE/blob/ec3f3f620a67df0a3e72f8e9227d09e4543dbb99/ipd_DiCE.py#L191
![noseed](https://user-images.githubusercontent.com/43922602/103469146-0c774980-4d87-11eb-950c-45f6ed420dce.png)

It shows that the results are reproducible only with seed = 42. But what if we do not know this? In such a case, I am not able to produce results consistent with the paper. And finding an appropriate seed (out of infinite integers) would make this implementation too sensitive to one hyperparameter (irrespective of the fact that it is theoretically sound). What if I want to implement the same algorithm in a different environment. The DiCE paper, however, shows IPD results across 5 runs converging to the average return of -1 (Page 8 of the DiCE paper).
Also, from where does the randomness comes? The parameters `theta` and `values` are always initialized to 0s. Rest all operations (other than `actions = m.sample()` in [line 80](https://github.com/alexis-jacq/LOLA_DiCE/blob/master/ipd_DiCE.py#L80)) are deterministic.

Thanks in advance!

**Comment by alexis-jacq** [comment]
Hi Videh Raj Nema, thanks for sharing these results. 
Indeed DiCE's paper results are highly dependent in the seed, and the algorithm only works "when it starts well". 
The problem is even worse with the initial LOLA algo (I could never reproduce that one), and was corrected by adding a smarter baseline in a more recent paper.
For the "seed=42", I guess I was lucky: I always start an implementation this number.

**Comment by vrn25** [comment]
Thanks, @alexis-jacq. The variance due to the Monte-Carlo rollouts is very high and I think using better advantage estimators can make the algorithm more robust. Unfortunately, the LOLA-DiCE objective, by its design considers only the full Monte-Carlo rollouts. But we can use [Loaded DiCE](https://arxiv.org/abs/1909.10549) - I found this recently. The objective function here requires just a few minimal changes to the original DiCE objective and lets us use critics.
I will try to submit a PR to include the code and the results :)

**Comment by waterhorse1** [comment]
Hi @vrn25, I am wondering have you make LOLA work with Loaded DICE.

**Comment by Aidandos** [comment]
Hello everyone! By increasing the batch size you can get consistent results over different seeds. I set the batch size to 8192 and ran it over 10 different seeds for look-ahead values of {1,2,3} and a **high** inner learning rate of 1.0. The shaded area denotes the standard error.
![og_dice](https://user-images.githubusercontent.com/8877292/157431119-1e2e928a-09db-4777-bd59-e400542ac1fa.png)

For Loaded Dice we get slightly better results, thought still not perfect:

![lola_dice](https://user-images.githubusercontent.com/8877292/157434346-a2d94e5e-bffe-453a-b122-0f09164a3980.png)
 
 
 I hope that helps!

**Comment by waterhorse1** [comment]
@Aidandos Indeed! Also refer to the second experiments in our new meta-gradient analysis paper  https://arxiv.org/abs/2112.15400 for systematically analysing what influences LOLA-DICE performance!
